README.md
# How to publish your own [neuron] site

[neuron] is a note-taking app optimized for publishing. Use this template repository to get started with [publishing](https://neuron.zettel.page/778816d3.html) your own neuron site that looks like [one of these][examples].

- Go to <https://github.com/srid/neuron-template/generate>
- Give your new repository a name, say `mynotes`
- Select "*Include all branches*" ([might be necessary to get the site to publish](https://stackoverflow.com/a/47368231/55246))
- Click "Create repository from template"
  - Note: if you are on the free GitHub plan, your repository should be public for GitHub to publish it.

GitHub will now build the site and serve it at: `https://<yourgithubusername>.github.io/mynotes/`.

For more information, see [neuron documentation][neuron] as well as the [GitHub Pages guide](https://help.github.com/en/github/working-with-github-pages).

## Set your site metadata

- In your new repository, edit the `neuron.dhall` file to set your site configuration (such as title, author, color theme) to suitable values.

## How to edit and add notes

Assuming you have changed the `editUrl` configuration in `neuron.dhall` (see the above section), you can simply click the "edit" icon on the published site to edit any note (see [Editing files in your repository](https://help.github.com/en/github/managing-files-in-a-repository/editing-files-in-your-repository) and [Creating new files](https://help.github.com/en/github/managing-files-in-a-repository/creating-new-files)). On every change, your site should eventually rebuild.

To understand how linking works, read [the neuron guide on Linking][linking].

For other ways to edit your notes (editors, web interface), see the [neuron guide][create]. In particular, [Cerveau](https://www.cerveau.app/) is the easiest way to edit your notes on the go.

Got questions? Checkout the [[faq]]. To find who else is using this template *publicly on GitHub*, [see here](https://github.com/search?o=desc&q=filename%3Aneuron.dhall&s=indexed&type=Code).

## Adding Custom Domain

Add `cname` option in `./github/workflows/publish.yaml` under
`peaceiris/actions-gh-pages@vX.X.X-X` as demonstrated in this
[pr](https://github.com/peaceiris/actions-gh-pages/pull/119)

[neuron]: https://neuron.zettel.page
[examples]: https://marklcrns.github.io/my-zettelkasten/
[linking]: https://neuron.zettel.page/linking
[create]: https://neuron.zettel.page/create



history.md
---
title: History
slug: history
tags:
  - history
---


``` {=html}
<script src="https://d3js.org/d3.v6.min.js"></script>
<link rel="stylesheet" href="./static/css/2d-graph.css">

<div id="2d-graph-container"></div>

<script src="./static/js/2d-graph.js" type="module"></script>
```

> Drag to pin a node. Ctrl + Click to unpin a node. Double click node to view
> note. Nodes are colored and sized based on incoming degree, so notes more
> heavily linked to will grow in relative size.

_Credits goes to Devon Morris' [Zettelkasten](https://devonmorris.dev/) for his
2D force-directed graph made with [D3.js](https://d3js.org/)_

[[z:zettels?tag=history/**&timeline&grouped]]


tag-add.sh
#!/usr/bin/env bash

set -o pipefail
set -o nounset

SCRIPT_PATH="$(realpath -- "${0}")"
while [ -h "${SCRIPT_PATH}" ]; do
  SCRIPT_DIR="$(cd -P "$(dirname "${SCRIPT_PATH}")" >/dev/null 2>&1 && pwd)"
  SCRIPT_PATH="$(readlink "${SCRIPT_PATH}")"
  [[ "${SCRIPT_PATH}" != /* ]] && SCRIPT_PATH="${SCRIPT_DIR}/${SCRIPT_PATH}"
done
SCRIPT_DIR="$(cd -P "$(dirname "${SCRIPT_PATH}")" >/dev/null 2>&1 && pwd)"
SCRIPT_NAME="$(basename -- "${SCRIPT_PATH}")"

usage() {
cat << EOF
USAGE: <required> [optional]

  ./${SCRIPT_NAME} [!] <content-pattern> <tag> [start] [pos] [directory]
  ./${SCRIPT_NAME} "lorem ipsum" 'new-tag'
  ./${SCRIPT_NAME} "lorem ipsum" 'new-tag' 1 .
  ./${SCRIPT_NAME} "lorem ipsum" 'new-tag' "old-tag"
  ./${SCRIPT_NAME} ! "lorem ipsum" 'new-tag' "old-*" test-dir/

  [!]                 Only include files not matching <content-pattern>.
  <content-pattern>   regex pattern to match the file(s) content to be included
                      in tag appending.
  <tag>               name of tag to append.
  [start]             0 (default) to append to start of tags meta list, 1 to
                      append at the end of the list, or a tag to append before
                      or after to. see [pos].
  [pos]               0 to append before or 1 (default) to append after tag from
                      the matching [start].
  [directory]         directory (default '.') to search into, recursively.

EOF
}

if [[ ${#} -gt 6 ]] || [[ ${#} -lt 2 ]]; then
  usage
  echo "ERROR: arg count: ${#}"
  exit 1
fi

if [[ -n ${1+x} ]]; then
  if [[ "${1}" == "!" ]]; then
    GREP_FLAGS=-rLi
    shift 1
  else
    GREP_FLAGS=-rli
  fi
else
  usage
  exit 1
fi

if [[ -n ${1+x} ]]; then
  PATTERN="${1}"
else
  usage
  exit 1
fi

if [[ -n ${2+x} ]]; then
  TAG="${2}"
else
  usage
  exit 1
fi

START="${3:-0}"
if ! [[ -d "${START}" ]] || [[ "${START}" -eq "${START}" ]]; then
  shift 1
fi

POS="${3:-1}"
if [[ "${POS}" != "." ]] || [[ "${POS}" -eq "${POS}" ]]; then
  shift 1
fi

TARGET_DIR="${3:-.}"
if ! [[ -d "${TARGET_DIR}" ]]; then
  usage
  echo "ERROR: Target directory does not exist"
  exit 1
else
  # Strip trailing whitespace
  TARGET_DIR=$(echo "${TARGET_DIR}" | sed 's,/*$,,')
fi

readonly GREP_FLAGS
readonly PATTERN
readonly TAG
readonly START
readonly POS
readonly TARGET_DIR

readonly SCRIPT_PATH
readonly SCRIPT_DIR
readonly SCRIPT_NAME
readonly INDENT="  "

MATCH_FILES="$( \
  grep ${GREP_FLAGS} \
  --exclude-dir=".git" \
  --exclude-dir=".neuron" \
  --exclude-dir=".github" \
  --exclude-dir=".static" \
  --exclude-dir=".vscode" \
  --exclude="index.md" \
  --exclude="todo.md" \
  --exclude="faq.md" \
  --exclude="README.md" \
  -- \
  "${PATTERN}" \
  ${TARGET_DIR}/*.md \
)"

for file in $MATCH_FILES; do
  begin_meta=0
  begin_tags=0
  line_nr=0

  # Append at the end of tags
  while IFS='' read -r line || [ -n "${line}" ]; do
    ((line_nr++))
    # Bound search only within metadata space
    if [[ "${line}" == "---" ]]; then
      if [[ ${begin_meta} -eq 0 ]]; then
        begin_meta=1
        continue
      else
        break
      fi
    fi

    # Bound search only under tags meta
    if [[ "${line}" =~ tags:* ]] && [[ ${begin_tags} -eq 0 ]]; then
      begin_tags=1
      if [[ ${START} == '1' ]]; then
        ((line_nr++))
        break
      fi
      continue
    elif [[ ${begin_tags} -eq 1 ]]; then
      if [[ "${line}" =~ ${INDENT}-\ ${START} ]]; then
        if [[ ${POS} -eq 1 ]]; then
          ((line_nr++))
        fi
        break
      elif [[ "${line}" =~ ${INDENT}-* ]]; then
        continue
      else
        break
      fi
    else
      continue
    fi
  done < ${file}

  # Additionally append two --- if not existing at beginning of the file
  if [[ ${begin_meta} -eq 0 ]]; then
    awk "NR==1{print \"---\n---\"}1" "${file}" > tmp && mv tmp "${file}"
    line_nr=2
  fi
  # Additionally append tags meta if not exists
  if [[ ${begin_tags} -eq 0 ]]; then
    awk "NR==${line_nr}{print \"tags:\"}1" "${file}" > tmp && mv tmp "${file}"
    ((line_nr++))
  fi

  echo "Appending '${TAG}' tag into ${file}..."
  awk "NR==${line_nr}{print \"${INDENT}- ${TAG}\"}1" "${file}" > tmp && mv tmp "${file}"
done

if [[ -z ${MATCH_FILES} ]]; then
  echo "no matching file"
else
  echo "Done!"
fi


tag-del.sh
#!/usr/bin/env bash

set -o pipefail
set -o nounset

SCRIPT_PATH="$(realpath -- "${0}")"
while [ -h "${SCRIPT_PATH}" ]; do
  SCRIPT_DIR="$(cd -P "$(dirname "${SCRIPT_PATH}")" >/dev/null 2>&1 && pwd)"
  SCRIPT_PATH="$(readlink "${SCRIPT_PATH}")"
  [[ "${SCRIPT_PATH}" != /* ]] && SCRIPT_PATH="${SCRIPT_DIR}/${SCRIPT_PATH}"
done
SCRIPT_DIR="$(cd -P "$(dirname "${SCRIPT_PATH}")" >/dev/null 2>&1 && pwd)"
SCRIPT_NAME="$(basename -- "${SCRIPT_PATH}")"

usage() {
cat << EOF
USAGE: <required> [optional]

  ./${SCRIPT_NAME} [!] <content-pattern> <tag> [all] [directory]
  ./${SCRIPT_NAME} "lorem ipsum" 'new-tag' .
  ./${SCRIPT_NAME} ! "lorem ipsum" 'new-*' test-dir/
  ./${SCRIPT_NAME} "lorem ipsum" 'new-*' 1
  ./${SCRIPT_NAME} ! "lorem ipsum" 'new-*' 1 test-dir/

  [!]                 Only include files not matching <content-pattern>.
  <content-pattern>   regex pattern to match the file(s) content to be included
                      in tag deletion.
  <tag>               name of tag to delete.
  [all]               0 (default) to delete first match. 1 to delete all
                      matching pattern.
  [directory]         directory (default '.') to search into, recursively.

EOF
}

if [[ ${#} -gt 5 ]] || [[ ${#} -lt 2 ]]; then
  usage
  echo "ERROR: arg count: ${#}"
  exit 1
fi

if [[ -n ${1+x} ]]; then
  if [[ "${1}" == "!" ]]; then
    GREP_FLAGS=-rLi
    shift 1
  else
    GREP_FLAGS=-rli
  fi
else
  usage
  exit 1
fi

if [[ -n ${1+x} ]]; then
  PATTERN="${1}"
else
  usage
  exit 1
fi

if [[ -n ${2+x} ]]; then
  TAG="${2}"
else
  usage
  exit 1
fi

ALL="${3:-0}"
if [[ "${ALL}" != "." ]] || [[ "${ALL}" -eq "${ALL}" ]]; then
  shift 1
fi

TARGET_DIR="${3:-.}"
if ! [[ -d "${TARGET_DIR}" ]]; then
  usage
  echo "ERROR: Target directory does not exist"
  exit 1
else
  # Strip trailing whitespace
  TARGET_DIR=$(echo "${TARGET_DIR}" | sed 's,/*$,,')
fi

readonly GREP_FLAGS
readonly PATTERN
readonly TAG
readonly ALL
readonly TARGET_DIR

readonly SCRIPT_PATH
readonly SCRIPT_DIR
readonly SCRIPT_NAME
readonly INDENT="  "

MATCH_FILES="$( \
  grep ${GREP_FLAGS} \
  --exclude-dir=".git" \
  --exclude-dir=".neuron" \
  --exclude-dir=".github" \
  --exclude-dir=".static" \
  --exclude-dir=".vscode" \
  --exclude="index.md" \
  --exclude="todo.md" \
  --exclude="faq.md" \
  --exclude="README.md" \
  -- \
  "${PATTERN}" \
  ${TARGET_DIR}/*.md \
)"

for file in $MATCH_FILES; do
  line_nr="$(awk "/tags:/{ print NR; exit }" "${file}")"

  begin_meta=0
  begin_tags=0
  while IFS='' read -r line || [ -n "${line}" ]; do

    # Bound search only within metadata space
    if [[ "${line}" == "---" ]]; then
      if [[ ${begin_meta} -eq 0 ]]; then
        begin_meta=1
        continue
      else
        break
      fi
    fi

    # Bound search only under tags meta
    if [[ "${line}" =~ tags:* ]] && [[ ${begin_tags} -eq 0 ]]; then
      begin_tags=1
      continue
    elif [[ ${begin_tags} -eq 1 ]]; then

      if [[ "${line}" =~ ${INDENT}-\ ${TAG} ]]; then
        ((line_nr++))
        echo "Deleting '${TAG}' tag from ${file}..."
        sed -i "${line_nr}d" "${file}"
        if [[ ${ALL} -eq 0 ]]; then
          break
        else
          ((line_nr--))
        fi
      elif [[ "${line}" =~ ${INDENT}-* ]]; then
        ((line_nr++))
        continue
      else
        break
      fi
    else
      continue
    fi
  done < ${file}

  # Delete tags: meta if no more tags
  if [[ ${begin_tags} -eq 1 ]]; then
    line_nr="$(awk "/tags:/{ print NR; exit }" "${file}")"
    next=$(sed -n "$((line_nr+1))p" "${file}")
    if [[ -n ${line_nr} ]] && ! [[ "${next}" =~ ${INDENT}-.* ]]; then
      sed -i "${line_nr}d" "${file}"
    fi
  fi

  # Delete both --- metadata fence if no more metas
  if [[ ${begin_meta} -eq 1 ]]; then
    line_nr="$(awk "/---/{ print NR; exit }" "${file}")"
    next=$(sed -n "$((line_nr+1))p" "${file}")
    if [[ -n ${line_nr} ]] && [[ "${next}" == '---' ]]; then
      sed -i "${line_nr}d" "${file}"
      sed -i "${line_nr}d" "${file}"
    fi
  fi
done

if [[ -z ${MATCH_FILES} ]]; then
  echo "no matching file"
else
  echo "Done!"
fi


static/graph.js
// 2D Directed Graph
d3.json("../cache.json", function (d) {
  return d;
})
  .then(function (data) {
    var graph = data.Graph;

    // Set in_degree to 0
    Object.values(graph.vertices).forEach((n) => {
      n.inDegree = 0;
    });

    // Compute number of incoming links
    var links = [];
    Object.entries(graph.adjacencyMap).forEach(([k,e]) => {
      Object.keys(e).forEach((n) => {
        graph.vertices[n].inDegree += 1;
        var link = {};
        link.source = k;
        link.target = n;
        links.push(link);
      })
    });

    // Create a scale for radius
    var radius_scale = d3
      .scaleSqrt()
      .domain([
        d3.min(Object.values(graph.vertices), (n) => n.inDegree),
        d3.max(Object.values(graph.vertices), (n) => n.inDegree),
      ])
      .range([3, 20]);

    // Create color scale
    var unpinned_color_scale = d3
      .scaleLinear()
      .domain([
        d3.min(Object.values(graph.vertices), (n) => n.inDegree),
        d3.max(Object.values(graph.vertices), (n) => n.inDegree),
      ])
      .range(["#ffffff", "#19a6a6"]);

    // Create color scale
    var pinned_color_scale = d3
      .scaleLinear()
      .domain([
        d3.min(Object.values(graph.vertices), (n) => n.inDegree),
        d3.max(Object.values(graph.vertices), (n) => n.inDegree),
      ])
      .range(["#ffffff", "#cc3d3d"]);


    var pandocDiv = document.getElementsByClassName("pandoc")[0];

    var width = pandocDiv.clientWidth;
    var height = 800;

    var force = d3
      .forceSimulation()
      .nodes(Object.values(graph.vertices))
      .force("link", d3.forceLink(links).distance(100).id((d) => d.ID))
      .force("center", d3.forceCenter(width / 2, height / 2))
      .force("x", d3.forceX())
      .force("y", d3.forceY())
      .force("charge", d3.forceManyBody().strength(-250).distanceMin(50))
      .alphaTarget(1)
      .on("tick", tick);


    var svg = d3
      .select(".pandoc")
      .insert("svg", ":nth-child(3)")
      .attr("width", width)
      .attr("height", height);

    svg
      .append('defs')
      .append('marker')
      .attr('id', 'arrow')
      .attr('viewBox', [0, 0, 10, 10])
      .attr('refX', 1)
      .attr('refY', 5)
      .attr('markerWidth', 10)
      .attr('markerHeight', 15)
      .attr('orient', 'auto-start-reverse')
      .append('path')
      .attr('d', d3.line()([[0,0], [10,5], [0,10]]))
      .attr('stroke', 'black');

    // add the links
    var path = svg
      .append("g")
      .selectAll("path")
      .data(links)
      .enter()
      .append("path")
      .attr("class", (d) => "link")
      .attr("marker-end", "url(#arrow)");

    // define the nodes
    var node = svg
      .selectAll(".node")
      .data(force.nodes())
      .enter()
      .append("g")
      .attr("class", "node")
      .call(
        d3
        .drag()
        .on("start", dragstarted)
        .on("drag", dragged)
        .on("end", dragended))
      .on("click", onClick)
      .on("dblclick", dblClick);

    // add the nodes
    node
      .append("circle")
      .attr("id", (d) => "circle-" + d.index)
      .attr("r", (d) => radius_scale(d.inDegree))
      .attr( "fill",
        (d) => unpinned_color_scale(d.inDegree)
      );

    // add node labels
    node
      .append("text")
      .attr("id", (d) => "label-" + d.ID)
      .attr("class", "node_labels")
      .attr("x", (d) => radius_scale(d.inDegree))
      .attr("y", (d) => -radius_scale(d.inDegree))
      .text((d) => d.Title);

    // add the curvy lines
    function tick() {
      path.attr("d", function (d) {
        var dx = d.target.x - d.source.x,
          dy = d.target.y - d.source.y,
          dr = Math.sqrt(dx * dx + dy * dy);
        return (
          "M" +
          d.source.x +
          "," +
          d.source.y +
          "A" +
          dr +
          "," +
          dr +
          " 0 0,1 " +
          d.target.x +
          "," +
          d.target.y
        );
      });

      node.attr("transform", function (d) {
        return "translate(" + d.x + "," + d.y + ")";
      });
    }

    function dragstarted(event, d) {
      if (!event.active) force.alphaTarget(0.3).restart();
      d.fx = d.x;
      d.fy = d.y;
    }

    function dragged(event, d) {
      d.fx = event.x;
      d.fy = event.y;
      d.fixed = true;
      node.select("#circle-" + d.index)
        .attr("fill", pinned_color_scale(d.inDegree))
    }

    function dragended(event, d) {
      if (!event.active) force.alphaTarget(0);
      if (d.fixed) {
        d.fx = d.x;
        d.fy = d.y;
      } else {
        d.fx = null;
        d.fy = null;
      }
    }

    function onClick(event, d) {
      if (d.fixed && event.ctrlKey) {
        d.fixed = false;
        node.select("#circle-" + d.index)
          .attr("fill", unpinned_color_scale(d.inDegree))
        d.fx = null;
        d.fy = null;
      }
    }

    function dblClick(event, d) {
      window.open("../" + d.Slug + ".html");
    }

  })
  .catch(function (error) {
    console.log(error);
  });



static/js/search.js
function autocomplete(inp, arr) {
  /*the autocomplete function takes two arguments,
  the text field element and an array of possible autocompleted values:*/
  var currentFocus;
  /*execute a function when someone writes in the text field:*/
  inp.addEventListener("input", function(e) {
      var a, b, i, val = this.value;
      /*close any already open lists of autocompleted values*/
      closeAllLists();
      if (!val) { return false;}
      currentFocus = -1;
      /*create a DIV element that will contain the items (values):*/
      a = document.createElement("DIV");
      a.setAttribute("id", this.id + "autocomplete-list");
      a.setAttribute("class", "autocomplete-items");
      /*append the DIV element as a child of the autocomplete container:*/
      this.parentNode.appendChild(a);
      /*for each item in the array...*/
      for (i = 0; i < arr.length; i++) {
        /*check if the item starts with the same letters as the text field value:*/
        if (arr[i].substr(0, val.length).toUpperCase() == val.toUpperCase()) {
          /*create a DIV element for each matching element:*/
          b = document.createElement("DIV");
          /*make the matching letters bold:*/
          b.innerHTML = "<strong>" + arr[i].substr(0, val.length) + "</strong>";
          b.innerHTML += arr[i].substr(val.length);
          /*insert a input field that will hold the current array item's value:*/
          b.innerHTML += "<input type='hidden' value='" + arr[i] + "'>";
          /*execute a function when someone clicks on the item value (DIV element):*/
          b.addEventListener("click", function(e) {
              /*insert the value for the autocomplete text field:*/
              inp.value = this.getElementsByTagName("input")[0].value;
              /*close the list of autocompleted values,
              (or any other open lists of autocompleted values:*/
              closeAllLists();
          });
          a.appendChild(b);
        }
      }
  });
  /*execute a function presses a key on the keyboard:*/
  inp.addEventListener("keydown", function(e) {
      var x = document.getElementById(this.id + "autocomplete-list");
      if (x) x = x.getElementsByTagName("div");
      if (e.keyCode == 40) {
        /*If the arrow DOWN key is pressed,
        increase the currentFocus variable:*/
        currentFocus++;
        /*and and make the current item more visible:*/
        addActive(x);
      } else if (e.keyCode == 38) { //up
        /*If the arrow UP key is pressed,
        decrease the currentFocus variable:*/
        currentFocus--;
        /*and and make the current item more visible:*/
        addActive(x);
      } else if (e.keyCode == 13) {
        /*If the ENTER key is pressed, prevent the form from being submitted,*/
        e.preventDefault();
        if (currentFocus > -1) {
          /*and simulate a click on the "active" item:*/
          if (x) x[currentFocus].click();
        }
      }
  });
  function addActive(x) {
    /*a function to classify an item as "active":*/
    if (!x) return false;
    /*start by removing the "active" class on all items:*/
    removeActive(x);
    if (currentFocus >= x.length) currentFocus = 0;
    if (currentFocus < 0) currentFocus = (x.length - 1);
    /*add class "autocomplete-active":*/
    x[currentFocus].classList.add("autocomplete-active");
  }
  function removeActive(x) {
    /*a function to remove the "active" class from all autocomplete items:*/
    for (var i = 0; i < x.length; i++) {
      x[i].classList.remove("autocomplete-active");
    }
  }
  function closeAllLists(elmnt) {
    /*close all autocomplete lists in the document,
    except the one passed as an argument:*/
    var x = document.getElementsByClassName("autocomplete-items");
    for (var i = 0; i < x.length; i++) {
      if (elmnt != x[i] && elmnt != inp) {
        x[i].parentNode.removeChild(x[i]);
      }
    }
  }
  /*execute a function when someone clicks in the document:*/
  document.addEventListener("click", function (e) {
      closeAllLists(e.target);
  });
}

var countries = ["Afghanistan","Albania","Algeria","Andorra","Angola","Anguilla","Antigua & Barbuda","Argentina","Armenia","Aruba","Australia","Austria","Azerbaijan","Bahamas","Bahrain","Bangladesh","Barbados","Belarus","Belgium","Belize","Benin","Bermuda","Bhutan","Bolivia","Bosnia & Herzegovina","Botswana","Brazil","British Virgin Islands","Brunei","Bulgaria","Burkina Faso","Burundi","Cambodia","Cameroon","Canada","Cape Verde","Cayman Islands","Central Arfrican Republic","Chad","Chile","China","Colombia","Congo","Cook Islands","Costa Rica","Cote D Ivoire","Croatia","Cuba","Curacao","Cyprus","Czech Republic","Denmark","Djibouti","Dominica","Dominican Republic","Ecuador","Egypt","El Salvador","Equatorial Guinea","Eritrea","Estonia","Ethiopia","Falkland Islands","Faroe Islands","Fiji","Finland","France","French Polynesia","French West Indies","Gabon","Gambia","Georgia","Germany","Ghana","Gibraltar","Greece","Greenland","Grenada","Guam","Guatemala","Guernsey","Guinea","Guinea Bissau","Guyana","Haiti","Honduras","Hong Kong","Hungary","Iceland","India","Indonesia","Iran","Iraq","Ireland","Isle of Man","Israel","Italy","Jamaica","Japan","Jersey","Jordan","Kazakhstan","Kenya","Kiribati","Kosovo","Kuwait","Kyrgyzstan","Laos","Latvia","Lebanon","Lesotho","Liberia","Libya","Liechtenstein","Lithuania","Luxembourg","Macau","Macedonia","Madagascar","Malawi","Malaysia","Maldives","Mali","Malta","Marshall Islands","Mauritania","Mauritius","Mexico","Micronesia","Moldova","Monaco","Mongolia","Montenegro","Montserrat","Morocco","Mozambique","Myanmar","Namibia","Nauro","Nepal","Netherlands","Netherlands Antilles","New Caledonia","New Zealand","Nicaragua","Niger","Nigeria","North Korea","Norway","Oman","Pakistan","Palau","Palestine","Panama","Papua New Guinea","Paraguay","Peru","Philippines","Poland","Portugal","Puerto Rico","Qatar","Reunion","Romania","Russia","Rwanda","Saint Pierre & Miquelon","Samoa","San Marino","Sao Tome and Principe","Saudi Arabia","Senegal","Serbia","Seychelles","Sierra Leone","Singapore","Slovakia","Slovenia","Solomon Islands","Somalia","South Africa","South Korea","South Sudan","Spain","Sri Lanka","St Kitts & Nevis","St Lucia","St Vincent","Sudan","Suriname","Swaziland","Sweden","Switzerland","Syria","Taiwan","Tajikistan","Tanzania","Thailand","Timor L'Este","Togo","Tonga","Trinidad & Tobago","Tunisia","Turkey","Turkmenistan","Turks & Caicos","Tuvalu","Uganda","Ukraine","United Arab Emirates","United Kingdom","United States of America","Uruguay","Uzbekistan","Vanuatu","Vatican City","Venezuela","Vietnam","Virgin Islands (US)","Yemen","Zambia","Zimbabwe"];




window.addEventListener("load", function(){
  $.getJSON('../cache.json', function(data) {
    var graph = fetchData(data.Graph).nodes
    console.log(graph);

    // let miniSearch = new MiniSearch({
    //   fields: ['title'], // fields to index for full-text search
    //   storeFields: ['title', 'date', 'tags'] // fields to return with search results
    // })
    // 
    // // Index all documents
    // miniSearch.addAll(graph)
    // let results = miniSearch.search('Reward Sens')
    // console.log(results);

    autocomplete(document.getElementById("myInput"), countries);
  })
});

function fetchData(data) {
  var graph = {};
  var excludedNodes = ["index", "README", "faq", "LICENSE"];

  // Build nodes
  var nodes = [];
  Object.entries(data.vertices).forEach(([node,e]) => {
    if (!excludedNodes.includes(e.ID)) {
      var node = {}
      // Meta
      node.id = e.ID;
      node.title = e.Title;
      node.slug = e.Slug;
      node.date = new Date(e.Date).toDateString();
      node.tags = e.Meta.tags;
      nodes.push(node);
    }
  })

  graph.nodes = nodes;
  return graph;
}


static/js/2d-graph.js
var excludedNodes = [
  "index",
  "zettel",
  "book",
  "programming",
  "science",
  "psychology",
  "history",
  "README",
  "faq",
  "LICENSE"
];

function excludeNonMatchingTags(graph, tags) {
  Object.values(graph.vertices).forEach((n) => {
    for (const tag of n.Meta.tags) {
      if (tags.includes(tag)) return;
    }
    if (!excludedNodes.includes(n.ID)) excludedNodes.push(n.ID);
  });
}

// 2D Directed Graph
d3.json("../cache.json", function (d) {
  return d;
})
  .then(function (data) {
    excludeNonMatchingTags(data.Graph, $('#neuron-tags-pane').text());
    var graph = buildGraph(data.Graph);

    var pandocDiv = document.getElementsByClassName("pandoc")[0];
    var width = pandocDiv.clientWidth;
    var height = 600;

    // Create a scale for radius
    var radius_scale = d3
      .scaleSqrt()
      .domain([
        d3.min(Object.values(graph.nodes), (n) => n.size),
        d3.max(Object.values(graph.nodes), (n) => n.size),
      ])
      .range([3, Math.max(width/45, 10)]);

    // Create color scale
    var unpinned_color_scale = d3
      .scaleLinear()
      .domain([
        d3.min(Object.values(graph.nodes), (n) => n.size),
        d3.max(Object.values(graph.nodes), (n) => n.size),
      ])
      .range(["#ffffff", "#cc3d3d"]);

    // Create color scale
    var pinned_color_scale = d3
      .scaleLinear()
      .domain([
        d3.min(Object.values(graph.nodes), (n) => n.clusterid),
        d3.max(Object.values(graph.nodes), (n) => n.clusterid),
      ])
      .range(["#ffffff", "#19a6a6"]);


    var force = d3
      .forceSimulation()
      .nodes(Object.values(graph.nodes))
      .force("link", d3.forceLink(graph.links).distance(70).id((d) => d.id))
      .force("center", d3.forceCenter(width / 2, height / 2))
      .force("x", d3.forceX())
      .force("y", d3.forceY())
      .force("charge", d3.forceManyBody().strength(-width/4).distanceMin(40))
      .alphaTarget(1)
      .on("tick", tick);


    var svg = d3
      .select(".pandoc")
      .insert("svg", ":nth-child(3)")
      .attr("width", width)
      .attr("height", height);

    svg
      .append('defs')
      .append('marker')
      .attr('id', 'arrow')
      .attr('viewBox', [0, 0, 10, 10])
      .attr('refX', 1)
      .attr('refY', 5)
      .attr('markerWidth', 10)
      .attr('markerHeight', 15)
      .attr('orient', 'auto-start-reverse')
      .append('path')
      .attr('d', d3.line()([[0,0], [10,5], [0,10]]))
      .attr('stroke', 'black');

    // add the links
    var path = svg
      .append("g")
      .selectAll("path")
      .data(graph.links)
      .enter()
      .append("path")
      .attr("class", (d) => "link")
      // .attr("marker-end", "url(#arrow)");

    // define the nodes
    var node = svg
      .selectAll(".node")
      .data(force.nodes())
      .enter()
      .append("g")
      .attr("class", "node")
      .call(
        d3
        .drag()
        .on("start", dragstarted)
        .on("drag", dragged)
        .on("end", dragended))
      .on("click", onClick)
      .on("dblclick", dblClick);

    // add the nodes
    node
      .append("circle")
      .attr("id", (d) => "circle-" + d.id)
      .attr("r", (d) => radius_scale(d.size))
      .attr( "fill",
        (d) => unpinned_color_scale(d.size)
      );

    // add node labels
    node
      .append("text")
      .attr("id", (d) => "label-" + d.id)
      .attr("class", "node_labels")
      .attr("x", (d) => radius_scale(d.size))
      .attr("y", (d) => -radius_scale(d.size))
      .text((d) => d.name);

    // add the curvy lines
    function tick() {
      path.attr("d", function (d) {
        var dx = d.target.x - d.source.x,
          dy = d.target.y - d.source.y,
          dr = Math.sqrt(dx * dx + dy * dy);
        return (
          "M" +
          d.source.x +
          "," +
          d.source.y +
          "A" +
          dr +
          "," +
          dr +
          " 0 0,1 " +
          d.target.x +
          "," +
          d.target.y
        );
      });

      node.attr("transform", function (d) {
        return "translate(" + d.x + "," + d.y + ")";
      });
    }

    function dragstarted(event, d) {
      if (!event.active) force.alphaTarget(0.3).restart();
      d.fx = d.x;
      d.fy = d.y;
    }

    function dragged(event, d) {
      d.fx = event.x;
      d.fy = event.y;
      d.fixed = true;
      node.select("#circle-" + d.id)
        .attr("fill", pinned_color_scale(d.clusterid))
    }

    function dragended(event, d) {
      if (!event.active) force.alphaTarget(0);
      if (d.fixed) {
        d.fx = d.x;
        d.fy = d.y;
      } else {
        d.fx = null;
        d.fy = null;
      }
    }

    function onClick(event, d) {
      if (d.fixed && event.ctrlKey) {
        d.fixed = false;
        node.select("#circle-" + d.id)
          .attr("fill", unpinned_color_scale(d.size))
        d.fx = null;
        d.fy = null;
      }
    }

    function dblClick(event, d) {
      window.open("../" + d.slug + ".html");
    }

  })
  .catch(function (error) {
    console.log(error);
  });

function buildGraph(data) {
  var graph = {};
  var nodeMap = {};
  var clusterCount = 0;

  // Build nodes
  var nodes = [];
  Object.entries(data.vertices).forEach(([node,e]) => {
    if (!excludedNodes.includes(e.ID)) {
      var node = {}
      // Meta
      node.id = e.ID;
      node.name = e.Title;
      node.slug = e.Slug;
      node.date = new Date(e.Date).toDateString();
      node.tags = e.Meta.tags;
      node.clusterid = 0;
      node.size = 0;
      // For cross-linking
      node.neighbors = [];
      node.links = [];
      // Push to nodeMap for ref
      nodeMap[node.id] = nodes.length;

      nodes.push(node);
    }
  })

  // Build links
  var links = [];
  Object.entries(data.adjacencyMap).forEach(([source, entry]) => {
    if (!$.isEmptyObject(entry)) {
      Object.entries(entry).forEach((e) => {
        var target = e[0];
        var folgeType = e[1][0];

        if (!excludedNodes.includes(source) && !excludedNodes.includes(target)) {
          var link = {};
          if (folgeType === "folgeinv") {
            link.source = target;
            link.target = source;
          } else if (folgeType === "folge") {
            link.source = source;
            link.target = target;
          } else {
            return;   // equivalent to conventional `continue`
          }

          // Inherit clusterid from source node
          if (nodes[nodeMap[link.source]].clusterid === 0) {
            nodes[nodeMap[link.source]].clusterid = ++clusterCount;
          }
          nodes[nodeMap[link.target]].clusterid = nodes[nodeMap[link.source]].clusterid;

          var visited = [];
          // recache clusterid from target's existing links
          recacheNodeImmediateLinks(link.target, links, nodeMap, visited, false);
          recacheNodeImmediateLinks(link.source, links, nodeMap, visited, false);
          // Simply resize source without changing clusterid
          recacheNodeImmediateLinks(link.source, links, nodeMap, visited, true);

          links.push(link);
        }
      });
    }
  })

  function recacheNodeImmediateLinks(source, links, nodeMap, visited, isResize) {
    for (var i = 0; i < links.length; ++i) {
      if (!visited.includes(i) && links[i].source === source) {
        var scid = nodes[nodeMap[source]].clusterid;
        visited.push(i);
        if (nodes[nodeMap[links[i].target]].clusterid === scid) {
          nodes[nodeMap[source]].size += 2;
        } else if (!isResize){
          nodes[nodeMap[links[i].target]].clusterid = scid;
        }
        recacheNodeImmediateLinks(links[i].source, links, nodeMap, visited, isResize);
      }
    }
  }

  graph.nodes = nodes;
  graph.links = links;

  // Cross-link node objects
  graph.links.forEach(link => {
    const a = graph.nodes[nodeMap[link.source]];
    const b = graph.nodes[nodeMap[link.target]];
    a.neighbors.push(b);
    b.neighbors.push(a);
    a.links.push(link);
    b.links.push(link);
  });

  return graph;
}

